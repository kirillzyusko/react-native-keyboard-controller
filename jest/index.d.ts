declare module 'react-native-keyboard-controller/jest' {
  function setKeyboardPosition(height: number);
  function setKeyboardVisibleHeight(height: number);
}
