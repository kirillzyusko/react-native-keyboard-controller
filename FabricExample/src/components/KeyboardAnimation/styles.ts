import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  box: {
    width: 60,
    height: 60,
    marginVertical: 20,
  },
  row: {
    flexDirection: 'row',
  },
  center: {
    justifyContent: 'center',
  },
});
