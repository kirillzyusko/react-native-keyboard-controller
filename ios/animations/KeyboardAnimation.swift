//
//  KeyboardAnimation.swift
//  react-native-keyboard-controller
//
//  Created by Kiryl Ziusko on 04/05/2024.
//

import Foundation

protocol KeyboardAnimationProtocol {
  var startTime: CFTimeInterval { get }
  var diff: Double { get }
  var isIncreasing: Bool { get }

  func valueAt(time: Double) -> Double
  func timingAt(value: Double) -> Double
  func deriative(t: Double, x0: Double) -> Double
}

public class KeyboardAnimation: KeyboardAnimationProtocol {
  weak var animation: CAMediaTiming?

  // constructor variables
  let fromValue: Double
  let toValue: Double
  let speed: Float
  let timestamp: CFTimeInterval

  init(fromValue: Double, toValue: Double, animation: CAMediaTiming) {
    self.fromValue = fromValue
    self.toValue = toValue
    self.animation = animation
    speed = animation.speed
    timestamp = CACurrentMediaTime()
  }

  // public getters
  var startTime: CFTimeInterval {
    // when concurrent animation happens, then `.beginTime` remains the same
    return max(animation?.beginTime ?? timestamp, timestamp)
  }

  var diff: Double {
    return ((animation?.beginTime ?? timestamp) - timestamp).truncatingRemainder(dividingBy: UIUtils.nextFrame)
  }

  var isIncreasing: Bool {
    return fromValue < toValue
  }

  func valueAt(time _: Double) -> Double {
    fatalError("Method is not implemented in abstract class!")
  }
    
    func deriative(t _: Double, x0 _: Double) -> Double {
        fatalError("Method is not implemented in abstract class!")
    }

  func timingAt(value: Double) -> Double {
    var lowerBound = 0.0
    var upperBound = 1.0 // Assuming 1 second is the max duration for simplicity
    let tolerance = 0.00001 // Define how precise you want to be
    var tGuess = 0.0

    // Check the direction of the animation
    let isIncreasing = isIncreasing

    while (upperBound - lowerBound) > tolerance {
      tGuess = (lowerBound + upperBound) / 2
      let currentValue = valueAt(time: tGuess / Double(speed))

      // Adjust the condition to account for the direction of animation
      if (currentValue < value && isIncreasing) || (currentValue > value && !isIncreasing) {
        lowerBound = tGuess
      } else {
        upperBound = tGuess
      }
    }

    return tGuess / Double(speed)
  }
    
  func inverse(value: Double) -> Double {
      let x0 = toValue - fromValue
              let epsilon = 1e-6
              let maxIterations = 100
      // TODO: duration / 2
              var t = 0.25 // Initial guess
      
      for _ in 0..<maxIterations {
          
          // Derivative computation
          let dy_dt = deriative(t: t, x0: x0)
          
          if abs(dy_dt) < epsilon {
              break // Avoid division by zero
          }
          
          var t_new = t - valueAt(time: t) / dy_dt
          
          // Ensure t_new stays positive
          if t_new < 0 {
              t_new = t / 2
          }
          
          if abs(t_new - t) < epsilon {
              t = t_new
              break
          }
          
          t = t_new
      }
      
      return t / Double(speed)
  }
}
