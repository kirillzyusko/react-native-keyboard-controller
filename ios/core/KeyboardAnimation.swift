//
//  KeyboardAnimation.swift
//  react-native-keyboard-controller
//
//  Created by Kiryl Ziusko on 04/05/2024.
//

import Foundation

protocol KeyboardAnimationProtocol {
  var startTime: CFTimeInterval { get }
  var diff: Double { get }
  var isIncreasing: Bool { get }

  func valueAt(time: Double) -> Double
  func timingAt(value: Double) -> Double
}

public class KeyboardAnimation: KeyboardAnimationProtocol {
  private weak var animation: CAMediaTiming?

  // constructor variables
  let fromValue: Double
  let toValue: Double
  let speed: Float
  let timestamp: CFTimeInterval

  init(fromValue: Double, toValue: Double, animation: CAMediaTiming) {
    self.fromValue = fromValue
    self.toValue = toValue
    self.animation = animation
    speed = animation.speed
    timestamp = CACurrentMediaTime()
  }

  // public getters
  var startTime: CFTimeInterval {
    // when concurrent animation happens, then `.beginTime` remains the same
    return max(animation?.beginTime ?? timestamp, timestamp)
  }

  var diff: Double {
    return ((animation?.beginTime ?? timestamp) - timestamp).truncatingRemainder(dividingBy: UIUtils.nextFrame)
  }

  var isIncreasing: Bool {
    return fromValue < toValue
  }

  func valueAt(time _: Double) -> Double {
    fatalError("Method is not implemented in abstract class!")
  }

  func timingAt(value: Double) -> Double {
    var lowerBound = 0.0
    var upperBound = 1.0 // Assuming 1 second is the max duration for simplicity
    let tolerance = 0.00001 // Define how precise you want to be
    var tGuess = 0.0

    // Check the direction of the animation
    let isIncreasing = isIncreasing

    while (upperBound - lowerBound) > tolerance {
      tGuess = (lowerBound + upperBound) / 2
      let currentValue = valueAt(time: tGuess / Double(speed))

      // Adjust the condition to account for the direction of animation
      if (currentValue < value && isIncreasing) || (currentValue > value && !isIncreasing) {
        lowerBound = tGuess
      } else {
        upperBound = tGuess
      }
    }

    return tGuess / Double(speed)
  }
}
