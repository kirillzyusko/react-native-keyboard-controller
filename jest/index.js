import { Animated, ScrollView, View } from "react-native";

const values = {
  animated: {
    progress: new Animated.Value(0),
    height: new Animated.Value(0),
  },
  reanimated: {
    progress: { value: 0 },
    height: { value: 0 },
  },
};
const focusedInput = {
  input: {
    value: {
      target: 1,
      layout: {
        x: 0,
        y: 0,
        width: 200,
        height: 40,
        absoluteX: 0,
        absoluteY: 100,
      },
    },
  },
};

const mock = {
  // hooks
  /// keyboard
  useKeyboardAnimation: jest.fn().mockReturnValue(values.animated),
  useReanimatedKeyboardAnimation: jest.fn().mockReturnValue(values.reanimated),
  useResizeMode: jest.fn(),
  useGenericKeyboardHandler: jest.fn(),
  useKeyboardHandler: jest.fn(),
  /// input
  useReanimatedFocusedInput: jest.fn().mockReturnValue(focusedInput),
  useFocusedInputHandler: jest.fn(),
  /// module
  useKeyboardController: jest
    .fn()
    .mockReturnValue({ setEnabled: jest.fn(), enabled: true }),
  // modules
  KeyboardController: {
    setInputMode: jest.fn(),
    setDefaultMode: jest.fn(),
    dismiss: jest.fn(),
  },
  KeyboardEvents: {
    addListener: jest.fn(() => ({ remove: jest.fn() })),
  },
  // views
  KeyboardControllerView: "KeyboardControllerView",
  KeyboardGestureArea: "KeyboardGestureArea",
  // providers
  KeyboardProvider: "KeyboardProvider",
  // components
  KeyboardStickyView: View,
  KeyboardAvoidingView: View,
  KeyboardAwareScrollView: ScrollView,
};

module.exports = mock;
