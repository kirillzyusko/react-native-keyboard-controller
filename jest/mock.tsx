const Animated = require('react-native/Libraries/Animated/Animated');

let keyboardVisibleHeight = 300;

const values = {
  animated: {
    progress: new Animated.Value(0),
    height: new Animated.Value(0),
  },
  reanimated: {
    progress: { value: 0 },
    height: { value: 0 },
  },
};

const setKeyboardVisibleHeight = (height: number) => {
  keyboardVisibleHeight = height;
};
const setKeyboardPosition = (height: number) => {
  const progress = height / keyboardVisibleHeight;

  values.animated.progress.setValue(progress);
  values.animated.height.setValue(height);
  values.reanimated.height.value = height;
  values.reanimated.progress.value = progress;
};

const mock = {
  // hooks
  useKeyboardAnimation: () => values.animated,
  useReanimatedKeyboardAnimation: () => values.reanimated,
  useKeyboardAnimationReplica: () => values.animated,
  useResizeMode: jest.fn(),
  useGenericKeyboardHandler: jest.fn(),
  useKeyboardHandler: jest.fn(),
  // modules
  KeyboardController: {
    setInputMode: jest.fn(),
    setDefaultMode: jest.fn(),
  },
  KeyboardEvents: {
    addListener: jest.fn(() => ({ remove: jest.fn() })),
  },
  // views
  KeyboardControllerView: () => <></>,
  // mocks
  setKeyboardVisibleHeight,
  setKeyboardPosition,
};

module.exports = mock;
