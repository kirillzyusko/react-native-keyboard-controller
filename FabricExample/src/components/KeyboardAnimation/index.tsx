import React from 'react';
import { Animated, TextInput, View } from 'react-native';
import {
  useKeyboardAnimation,
  useKeyboardAnimationReplica,
} from 'react-native-keyboard-controller';
import styles from './styles';

export default function KeyboardAnimation() {
  const { height, progress } = useKeyboardAnimation();
  const { height: heightReplica } = useKeyboardAnimationReplica();

  return (
    <View style={styles.container}>
      <View>
        <Animated.View
          style={{
            width: 50,
            height: 50,
            backgroundColor: 'green',
            borderRadius: 25,
            transform: [
              {
                translateX: progress.interpolate({
                  inputRange: [0, 1],
                  outputRange: [0, 100],
                }),
              },
            ],
          }}
        />
      </View>
      <View>
        <TextInput
          style={{
            width: 200,
            marginTop: 50,
            height: 50,
            backgroundColor: 'yellow',
          }}
        />
        <View style={[styles.row, styles.center]}>
          <Animated.View
            style={{
              width: 50,
              height: 50,
              backgroundColor: 'red',
              borderRadius: 25,
              transform: [{ translateY: height }],
            }}
          />
          <Animated.View
            style={{
              width: 50,
              height: 50,
              backgroundColor: 'blue',
              borderRadius: 25,
              transform: [{ translateY: heightReplica }],
            }}
          />
        </View>
      </View>
    </View>
  );
}
