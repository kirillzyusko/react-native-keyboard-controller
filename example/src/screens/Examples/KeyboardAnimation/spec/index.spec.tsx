import React from 'react';
import { render } from '@testing-library/react-native';

import { setKeyboardPosition } from 'react-native-keyboard-controller/jest/mock';

import KeyboardAnimation from '../index';

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

describe('unit test sample', () => {
  it('should match to snapshot', () => {
    const screen = render(<KeyboardAnimation />);

    setKeyboardPosition(100);

    expect(screen).toMatchSnapshot();
  });
});
