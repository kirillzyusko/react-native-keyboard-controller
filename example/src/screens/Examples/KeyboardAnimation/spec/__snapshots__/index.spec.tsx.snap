// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit test sample should match to snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <View>
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": "green",
          "borderRadius": 25,
          "height": 50,
          "transform": Array [
            Object {
              "translateX": 33.33333333333333,
            },
          ],
          "width": 50,
        }
      }
    />
  </View>
  <View>
    <TextInput
      style={
        Object {
          "backgroundColor": "yellow",
          "height": 50,
          "marginTop": 50,
          "width": 200,
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          Object {
            "justifyContent": "center",
          },
        ]
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "red",
            "borderRadius": 25,
            "height": 50,
            "transform": Array [
              Object {
                "translateY": 100,
              },
            ],
            "width": 50,
          }
        }
      />
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "blue",
            "borderRadius": 25,
            "height": 50,
            "transform": Array [
              Object {
                "translateY": 100,
              },
            ],
            "width": 50,
          }
        }
      />
    </View>
  </View>
</View>
`;
